server/s3_post_multipartobject_action.cc

```diff
@@ -171,7 +171,7 @@ void S3PostMultipartObjectAction::parse_x_amz_tagging_header(
     for (struct evkeyval* header = key_value.tqh_first; header;
          header = header->next.tqe_next) {
 
-      decoded_key = evhttp_decode_uri(header->key);
+      decoded_key = evhttp_uridecode(header->key, 0, NULL);
```

Comment:
"The returned string must be freed by the caller."
Did you forget free() ?

Reply:
No, decoded_key is allocated memory from heap and is used by :
`new_object_tags_map[decoded_key] = header->value;`
This new_object_tags_map is used by validate_object_xml_tags and lives throughout the duration of the program, so we cannot free it. I tried freeing it which was causing failures.

Reply:
It's strange. A compiler must create temporary std::string object for the expression new_object_tags_map[decoded_key].
And decoded_key variable must be useless just after the expression.
Can you please retest

Reply:
new_object_tags_map is a std::map, not a temporary string, and decoded_key is one of its mapped entry.
I did testing couple of times and saw failures every time I tried to free decoded_key.

Reply:
Done, freeing up after the expression works.